lib/srv/db/spanner/grpcserver.go

```diff
@@ -0,0 +1,467 @@
+	// reason someone would want to use the full service account name, like for
+	// a service account in a different project?
+	dbUser := databaseUserToGCPServiceAccount(e.sessionCtx)
+	ts, err := e.Auth.GetSpannerTokenSource(ctx, e.sessionCtx.WithUser(dbUser))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// The context passed to DialPool controls dialing, but more importantly it
+	// controls GCP oauth2 credential refreshing.
+	// We do not want credential refreshing to stop working after the first
+	// oauth token expires (by default after one hour).
+	// Therefore use the client connection context rather than the RPC context.
+	cc, err := gtransport.DialPool(e.clientConnContext,
+		append(spannerapi.DefaultClientOptions(),
+			option.WithGRPCDialOption(grpc.WithTransportCredentials(credentials.NewTLS(tlsCfg))),
+			option.WithTokenSource(ts),
+			option.WithEndpoint(e.sessionCtx.Database.GetURI()),
+			// pool size seems to be adjusted to number of gRPC channels, which is
+			// 4 by default in Google's code and seemingly other clients as well,
+			// but with Teleport in the middle each downstream client connection
+			// will be handled by a separate engine.
+			// Therefore, avoid amplifying the number of connections dialed to
+			// GCP and just use a pool size of 1.
+			option.WithGRPCConnectionPool(1))...,
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	e.gcloudClient = cc
+	return spannerpb.NewSpannerClient(e.gcloudClient), trace.Wrap(err)
+}
+
+func (e *Engine) authorizeRPCRequest(ctx context.Context, targetID, procedure string, req any, skipAudit bool) error {
+	dbName, authzErr, checkedAccess := e.maybeStartSession(ctx, targetID)
+	if e.sessionStartErr != nil {
+		// don't emit any other audit events after a failed session start
+		return trace.Wrap(e.sessionStartErr)
+	}
+
+	if !checkedAccess {
+		dbName, authzErr = e.checkAccess(ctx, targetID)
+	}
+	if authzErr != nil {
+		e.closeOnce.Do(func() {
+			go e.Close()
+		})
+	}
+
+	if skipAudit {
+		return trace.Wrap(authzErr)
+	}
+
+	args, conversionErr := rpcRequestToStruct(req)
+	r := rpcInfo{
+		database:  dbName,
+		procedure: procedure,
+		args:      args,
+		err:       trace.NewAggregate(authzErr, conversionErr),
+	}
+	auditRPC(e.Context, e.Audit, e.sessionCtx, r)
+	return trace.Wrap(r.err)
+}
+
+// maybeStartSession checks access and starts a session, if one has not been
+// started already, and returns whether it checked access and started the session.
+func (e *Engine) maybeStartSession(ctx context.Context, targetID string) (string, error, bool) {
+	// check if we can avoid locking
+	if e.sessionStarted.Load() {
+		return "", nil, false
+	}
+
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	// check if the session was started while we were waiting for the lock
+	if e.sessionStarted.Load() {
+		return "", nil, false
+	}
+
+	dbName, authzErr := e.checkAccess(ctx, targetID)
+	if authzErr != nil {
+		go e.Close()
```

Comment:
we're calling it from a function invoked by a gRPC handler.

`e.Close` shuts down the gRPC server, blocks any further RPCs, and then waits for the RPCs that were already being handled to complete, so we have shutdown async to avoid waiting in a cycle.

Reply:
attempted to simplify in 5c3ef8d9cfe3b6024ae9bde6d696b4d1a9931288 which now uses a context cancel func instead of calling close in a goroutine